File: LeetCodeProblemsCMake/Problems/LeetCode873LengthOfLongestFibonacciSubsequence.cpp
//
// Created by saeed on 2/27/25.
//

#include "LeetCode873LengthOfLongestFibonacciSubsequence.h"

#include <iostream>
#include <ostream>
#include <unordered_map>
#include <unordered_set>



int LeetCode873LengthOfLongestFibonacciSubsequence::lenLongestFibSubseq(std::vector<int> &arr) {
    int n = arr.size();
    std::unordered_map<int, int> index;
    std::vector dp(n, std::vector(n, 2));

    int maxLen = 0;

    for (int i = 0; i < n; i++) {
        index[arr[i]] = i;
        for (int j = 0; j < i; j++) {
            int prev = arr[i] - arr[j];
            if (index.count(prev) && index[prev] < j) {
                int k = index[prev];
                dp[j][i] = dp[k][j] + 1;
                maxLen = std::max(maxLen, dp[j][i]);
            }
        }
    }

    return maxLen >= 3 ? maxLen : 0;
}


void LeetCode873LengthOfLongestFibonacciSubsequence::Run() {
    std::vector array =
            {2, 4, 5, 6, 7, 8, 11, 13, 14, 15, 21, 22, 34};
    // {1,3,7,11,12,14,18};
    // {1, 2, 3, 4, 5, 6, 7, 8};
    std::cout << lenLongestFibSubseq(array) << std::endl;
}

File: LeetCodeProblems/Problems/LeetCode873LengthOfLongestFibonacciSubsequence.cs
namespace LeetCodeProblems.Problems;

public class LeetCode873LengthOfLongestFibonacciSubsequence
{
    public int LenLongestFibSubseq(int[] arr) {
        int length=arr.Length;
        int[,] dp=new int[length,length];
        Dictionary<int,int> reverseIndices=new Dictionary<int,int>();
        for (int i = 0; i < length; i++)
        {
            reverseIndices.Add(arr[i],i);
        }

        int maxLength = 0;
        for (int i = 0; i < length; i++)
        {
            for (int j =0; j < i; j++)
            {
                if(!reverseIndices.ContainsKey(arr[i]-arr[j]))
                    continue;
                int index=reverseIndices[arr[i]-arr[j]];
                if(index>=j)
                    continue;
                dp[j,i] = dp[index,j]>0?dp[index,j]+1:3;
                if(dp[j,i]>maxLength)
                    maxLength=dp[j,i];
            }
        }
        return maxLength>=3?maxLength : 0;
    }
}